### PR TITLE
Non-selected checkbox questions are counted as answered

### PR DESCRIPTION
**Pull request by syntapy**
### Description
When an applicant does not select any items in a checkbox question, it counts as being answered

### Checklist
- [x] Modify browser test that skips a checkbox question

### Issue(s)
#931

### CODE REVIEW

File: browser-test/src/application_review.test.ts [hunk 1]
```diff
@@ -74,18 +79,23 @@ describe('normal application flow', () => {
...
     await applicantQuestions.answerFileUploadQuestion('file key');
     await applicantQuestions.clickNext();
 
-    // Applicant submits answers from review page.
+    // fill 4th application block. 
+    // skip one checkbox question.
+    await applicantQuestions.answerCheckboxQuestion(['clowns']);
+    await applicantQuestions.clickNext();
+
+    // submit
```
Comment (syntapy) [inline browser-test/src/application_review.test.ts:98]: Do I need to modify the test to verify that if it skips a non-checkbox question that it results in a review page with 'continue' on it? i.e. to make sure that me adding stuff didn't make it artificially result in 'submit'

**Review by carolinedanzi — COMMENTED** [review]
Hi Sami! Nice work on this, it's a tricky issue! While reviewing your PR I noticed it looks like someone else has added logic in `ApplicantServiceImpl#stageNormalUpdates` since the last time we talked that partially addresses this issue - basically, it adds an empty array to "clear" the existing data before updating. However, it doesn't quite work for checkbox questions - if I don't answer it, it won't let me submit. But if I go back to edit the answer and still don't answer the checkbox question, it does register as "answered" and then lets me submit. It's a weird issue - mind looking into this and updating so that it considers it "answered" the first time I don't answer? 

I like that you added browser tests - let's definitely keep those. To build off the newest code, we should leverage the new hidden input field added in `CheckboxQuestionRenderer`. Are you willing to update your approach?

Happy to chat in open office hours Friday!

**Comment by syntapy** [comment]
@carolinedanzi Oh yeah that makes sense. IYeah doing similar things in two different places (without the change in approach) could turn ugly 6-months from now when someone's trying to change something or whatnot...that is true in general, no?

I had noticed that other behavior you mentioned but it didn't register as something I should pay attention to

**Comment by carolinedanzi** [comment]
> @carolinedanzi Oh yeah that makes sense. Yeah doing similar things in two different places (without the change in approach) could turn ugly 6-months from now when someone's trying to change something or whatnot...that is true in general, no?
> 
> I had noticed that other behavior you mentioned but it didn't register as something I should pay attention to

Yes, great point! Maintenance is harder if we have multiple related special cases. Thank you for being willing to modify your approach and reduce the maintenance burden!

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java [hunk 2]
```diff
@@ -233,11 +233,13 @@ public void maybeClearArray(Path path) {
 
   private void putAt(Path path, Object value) {
     checkLocked();
+    putParentIfMissing(path);
     jsonData.put(path.parentPath().toString(), path.keyName(), value);
   }
 
   private void addAt(Path path, Object value) {
     checkLocked();
+    putParentIfMissing(path);
```
Comment (syntapy) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:242]: Should I add this one too? Not relevent to the checkbox thing I guess but seemed right based on concistency

Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:242]: Good catch, I think that is correct. Thanks for updating!

**Review by carolinedanzi — APPROVED** [review]

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java (hunk 2 above)
Comment (hmhsu) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:242]: Hey, this looks good. I have one request though, reading the `putParentIfMissing` function, there is a desire to simplify the function and remove the recursion. These two functions are called by the `putParentIfMissing` function and calling `putParentIfMissing` here further complicates the recursion tree. Would you instead add the function call in `maybeClearArray` directly before `putAt`? I think that will work, too, and leave a less complicated codebase for anyone who aspires to remove the recursion in the future.

Comment (syntapy) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java]: Ok I did it

Comment (hmhsu) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java]: Thanks!

**Review by hmhsu — APPROVED** [review]

**Pull request merged**